### PR TITLE
Fix ITelemetryModule singleton registration to support presence of keyed services

### LINKED ISSUE #2879: App Insights breaks when a keyed service is added
See the following bug report:

https://github.com/dotnet/extensions/issues/5222

When a keyed service is added, app insights breaks silently (no logging is sent to app insights) or breaks with an exception when the user tries to retrieve a telemetryclient.

Comments that lead to this issue being related to App Insights:

---------------
@kevincathcart-cas

ApplicationInsights has a AddSingletonIfNotExists method it uses internally that does not work correctly if there are any keyed services registered. And AddStandardResilienceHandler does register at least one keyed service.

ApplicationInsights is generally designed as best effort observability, which means that if it encounters errors it is not supposed to take down your service, and this is taken to the point where it mostly silently catches exceptions that occur inside .AddApplicationInsightsTelemetryWorkerService() (although it does call WorkerServiceEventSource.Instance.LogError(...) with the exception).

(This also means that doing .GetRequiredService<TelemetryClient>() is dangerous, as it is very deliberately the case that the registration could be missing if there were problems when trying to register the service.)

All that said, this is fundamentally just a bug with Application Insight's AddSingletonIfNotExists method, which really should be updated able to handle a ServiceCollection that contains keyed services, since that is a standard DI feature now, and users should not need to defer registering such services until after calling AddApplicationInsightsTelemetryWorkerService.

------------------
@joperezr

Thanks for the info @kevincathcart-cas, I actually didn't know that so today I learned 😃. I was able to confirm this by swapping the call to AddStandardResilienceHandler in your repro to a simple builder.Services.AddKeyedSingleton<MyService>("myService"); and that still repros if called before adding AppInsights. Given this isn't really an issue with the our resilience library, I'll go ahead and close this issue, and I'd suggest you opening one in the App Insights repo so that they can fix this issue.

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #2879
Relates to https://github.com/dotnet/extensions/issues/5222

## Changes

* Remove `AddSingletonIfNotExists` custom implementation for detection of registered services and call `TryAddEnumerable` instead because it understands keyed services added in 8.0.0

### CODE REVIEW

File: NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddSingletonTests.cs [hunk 1]
```diff
@@ -1,96 +0,0 @@
...
-using Xunit;
-
-namespace Microsoft.ApplicationInsights.AspNetCore.Tests.Extensions.ApplicationInsightsExtensionsTests
-{
-    /// <summary>
-    /// When working with IServiceCossection, it can store three types of Implementations:
-    /// ImplementationFactory, ImplementationInstance, and ImplementationType.
-    /// We want to be able to add a Singleton but only if a user hasn't already done so.
-    /// This class is to test all the various edge cases.
-    /// </summary>
```
Comment (CodeBlanch) [inline NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddSingletonTests.cs:21]: The issue is as of 8.0.0 this summary is false.

There are new "keyed" forms of the possible registration types:

image

`AddSingletonIfNotExists` could be made smarter to understand the new shape. But that would require an `8.0.0` bump for M.E.DI/.A in `Microsoft.ApplicationInsights.AspNetCore`.

Easier IMO to just call into `TryAddSingleton` and let it sort out what to do!

**Comment by CodeBlanch** [comment]
Sorry about the whitespace on diff! Blame VS 🤣

File: NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddApplicationInsightsTelemetryTests.cs [hunk 2]
```diff
@@ -37,6 +37,25 @@ namespace Microsoft.Extensions.DependencyInjection.Test
 
     public class AddApplicationInsightsTelemetryTests : BaseTestClass
     {
+        [Fact]
+        public static void TelemetryModuleResolvableWhenKeyedServiceRegistered()
```
Comment (TimothyMothra) [inline NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddApplicationInsightsTelemetryTests.cs:41]: This test appears to pass even without your change.
I copied it locally and ran it in main branch.

**Comment by xiang17** [comment]
The bug report is for .NET 8, but does it apply for other .NET versions and .NET Framework?

**Comment by CodeBlanch** [comment]
> The bug report is for .NET 8, but does it apply for other .NET versions and .NET Framework?

Could impact all runtimes. All user needs to do is take a dependency on _something_ which a) brings in M.E.DI 8.0.0+ and b) registers a keyed service. In the case of the bug report, it is actually `Polly` registering the keyed service.

File: NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddApplicationInsightsTelemetryTests.cs (hunk 2 above)
Comment (CodeBlanch) [reply NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddApplicationInsightsTelemetryTests.cs]: Interesting. It looks like without the fix, worker service can't find any registered `ITelemetryHttpModule` and the test fails. AspNetCore though finds `RequestTrackingTelemetryModule` registered so the test passes.

With the fix in place there are actually 7 `ITelemetryHttpModule`s registered. I updated the tests to validate all 7 are found when keyed service is added to the mix.

**Review by TimothyMothra — APPROVED** [review]
LGTM

**Review by rajkumar-rangaraj — APPROVED** [review]
LGTM, Thanks for fixing this quicker. Could you please add a changelog entry.

File: NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddApplicationInsightsTelemetryTests.cs (hunk 2 above)
Comment (CodeBlanch) [reply NETCORE/test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests/AddApplicationInsightsTelemetryTests.cs]: Just pushed updates to the test to improve them further. Added validation that a) `DiagnosticsTelemetryModule` is present in the final list and b) if the user supplied one we don't end up with 2.

**Comment by CodeBlanch** [comment]
@TimothyMothra @rajkumar-rangaraj I pushed a change to call `TryAddEnumerable` instead of `TryAddSingleton`. Please take another look.

I was looking through the history and trying to figure out why we had `AddSingletonIfNotExists` at all when there are other spots calling `TryAddSingleton`. What I came to realize is we want multiple `ITelemetryModule`s to be registered. But we only want 1 `DiagnosticsTelemetryModule`. `TryAddSingleton<ITelemetryModule, DiagnosticsTelemetryModule>()` won't add `DiagnosticsTelemetryModule` if some other `ITelemetryModule` is registered. That is incorrect! `TryAddEnumerable` is what we want here. What it will do is only add the `DiagnosticsTelemetryModule` registration if it hasn't already been registered.

So my guess is we had `AddSingletonIfNotExists` because `TryAddSingleton` didn't work and we didn't realize `TryAddEnumerable` was the thing to use 😄

File: NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs [hunk 3]
```diff
@@ -332,7 +312,7 @@ private static void AddCommonInitializers(IServiceCollection services)
         private static void AddCommonTelemetryModules(IServiceCollection services)
         {
             // Previously users were encouraged to manually add the DiagnosticsTelemetryModule.
-            services.AddSingletonIfNotExists<ITelemetryModule, DiagnosticsTelemetryModule>();
+            services.TryAddEnumerable(ServiceDescriptor.Singleton<ITelemetryModule, DiagnosticsTelemetryModule>());
```
Comment (CodeBlanch) [inline NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs:315]: @TimothyMothra FYI I didn't make any other changes here but really all of the `services.AddSingleton<ITelemetryModule, ...>();` calls should be `TryAddEnumerable` calls. If user only calls registration once everything will be fine. But if they call one of the "Add" extensions more than once, they'll get duplicates of all of these modules. Typically "Add" extensions should be safe to be called more than once which is accomplished by calling `TryAddSingleton` \ `TryAddEnumerable` 🤷

**Review by TimothyMothra — APPROVED** [review]

File: NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs (hunk 3 above)
Comment (TimothyMothra) [reply NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs:315]: I don't remember the full history, but I reviewed all my previous commit comments.

It looks like I was trying to add safety checks for every possible way that a user could have added the DiagnosticsTelemetryModule ([link 1](https://github.com/microsoft/ApplicationInsights-dotnet/pull/1886/commits/8a512f7323c28957003f746a9fe5fd19c40de205))([link 2](https://github.com/microsoft/ApplicationInsights-dotnet/pull/1886/commits/058008653cd81d6f3e9bd2401abba97ed8f1fcb1)). The TryAdd is substantially better :)

Thank you

**Comment by pcn-michaelulloa** [comment]
Any updates on this? 👀👀

File: NETCORE/test/Microsoft.ApplicationInsights.WorkerService.Tests/Microsoft.ApplicationInsights.WorkerService.Tests.csproj [hunk 4]
```diff
@@ -21,6 +21,7 @@
   <ItemGroup>
     <PackageReference Include="System.Net.Http" Version="4.3.4" />
     <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.1.0" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0" />
     <PackageReference Include="Microsoft.Extensions.Hosting" Version="2.1.0" />
     <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="2.1.0" />
```
Comment (RussKie) [inline NETCORE/test/Microsoft.ApplicationInsights.WorkerService.Tests/Microsoft.ApplicationInsights.WorkerService.Tests.csproj:26]: This looks as though it introduces version-mismatch. It's worth considering either using 2.1.0 across the board or bump all to 8.0.

**Comment by asvishnyakov** [comment]
@cijothomas @TimothyMothra Any news on that?

**Comment by ayuksekkaya** [comment]
Can this PR be merged please? There are many comments here complaining that this is causing issues for many people.

File: NETCORE/test/Microsoft.ApplicationInsights.WorkerService.Tests/Microsoft.ApplicationInsights.WorkerService.Tests.csproj (hunk 4 above)
Comment (rajkumar-rangaraj) [reply NETCORE/test/Microsoft.ApplicationInsights.WorkerService.Tests/Microsoft.ApplicationInsights.WorkerService.Tests.csproj:26]: This is a test project, should cause any issues.

**Review by rajkumar-rangaraj — APPROVED** [review]

**Comment by angularsen** [comment]
Related: https://github.com/microsoft/ApplicationInsights-dotnet/issues/2852

**Comment by TimothyMothra** [comment]
Just released 2.23.0-beta1 which contains this fix https://www.nuget.org/packages/Microsoft.ApplicationInsights.AspNetCore/2.23.0-beta1
